docs/developers/bitesize/workflowParsing/whatIsWom.md

```diff
@@ -0,0 +1,92 @@
+
+Examine the workflow below, making note of how the outputs of early calls become inputs for later calls. The first example is the `numbers` output of the `mkFile` call serving as the `in_file` input to the `grep` call. 
+```
+version 1.0
+
+##
+# Checks a simple branch and join operation.
+# We start with a task, branch into two parallel executions, and then rejoin to calculate the result.
+##
+
+workflow forkjoin {
+  call mkFile
+
+  call grep { input: in_file = mkFile.numbers }
+  call wc { input: in_file=mkFile.numbers }
+
+  call join { input: wcCount = wc.count, grepCount = grep.count }
+
+  output {
+    Int joined_proportion = join.proportion
+  }
+}
+
+task mkFile {
+  command <<<
+    for i in `seq 1 1000`
+    do
+      echo $i
+    done
+  >>>
+  output {
+    File numbers = stdout()
+  }
+  runtime {docker: "ubuntu:latest"}
+}
+
+task grep {
+  input {
+    String pattern
+    File in_file
+  }
+  command <<<
+    [ -f "~{in_file}" ] && grep '~{pattern}' ~{in_file} | wc -l
+  >>>
+  output {
+    Int count = read_int(stdout())
+  }
+  runtime {docker: "ubuntu:latest"}
+}
+
+task wc {
+  input {
+    File in_file
+  }
+  command <<<
+    [ -f "~{in_file}" ] && cat ~{in_file} | wc -l
+  >>>
+  output {
+    Int count = read_int(stdout())
+  }
+  runtime {docker: "ubuntu:latest"}
+}
+
+task join {
+  input {
+    Int grepCount
+    Int wcCount
+  }
+  command <<<
+    expr ~{wcCount} / ~{grepCount}
+  >>>
+  output {
+    Int proportion = read_int(stdout())
+  }
+  runtime {docker: "ubuntu:latest"}
+}
+```  
+
+Now, compare the workflow source to its WOM graph (generated with the `womtool womgraph` command).
+
+![Graph of forkjoin](forkjoin_graph.svg)
+
+Input values are ovals, outputs are hexagons. You can see that a hexagon in one node (call) becomes an oval in the next. 
+
+The the `grep` call is special because one of its two inputs, `pattern`, is not specified by any previous call within the bounds of the graph. This causes Cromwell to generate a blue "external graph input node" (`wom.graph.ExternalGraphInputNode` if you're looking at the code). Its value must be specified by the user in the inputs file of the workflow under key `grep.pattern`, and Cromwell will pass it into the `grep` call.   
+
+Finally, the `proportion` output of the `join` call is piped out of the bounds of the workflow graph by becoming the green `joined_proportion` graph output node (`wom.graph.ExpressionBasedGraphOutputNode` in the code). 
+
+Graph nodes inherit from trait `wom.graph.GraphNode`. The graph is constructed in class `wdl.transforms.base.wdlom2wom.WorkflowDefinitionElementToWomWorkflowDefinition`. The `convertGraphElements` function is especially interesting; it accepts a `Set` of nodes, generates the edges, links everything together, and spits out the finished graph as a `wom.graph.Graph`.
```

Comment:
> it accepts a `Set` of nodes

Depending on what you mean by "node".....

The inputs to the function are actually language elements (ie fragments of WDL file) which are converted into WOM nodes, complete with edges. IIRC you can't create a WOM node without specifying its inbound edges.

Reply:
When I say node I mean the contents of `wdl.transforms.base.wdlom2wom.WorkflowDefinitionElementToWomWorkflowDefinition.GraphLikeConvertInputs` - it seems like the `WorkflowGraphElement` member of that class contains the node information and the information needed to create edges (inputs and outputs) but not an actual list of edges

Of course you probably wrote most of this code so I will defer to you if the above is wrong

Reply:
What I mean is, the inputs are _conceptual_ nodes of a _conceptual_ graph, but they're not constructed-but-disconnected `wom.graph.GraphNode`s waiting to be joined together into a `wom.graph.Graph`.

Reply:
I might be being needlessly pedantic...

Reply:
hmm as a non-Redwood alum that seems like an important distinction, though I don't have a specific suggestion for wording that

Reply:
Updated. Somehow misread the class definition of the inputs.